geo/src/algorithm/relate/geomgraph/geometry_graph.rs

```diff
@@ -0,0 +1,387 @@
+                for line_string in &multi_line_string.0 {
+                    self.add_line_string(line_string);
+                }
+            }
+            GeometryCow::GeometryCollection(geometry_collection) => {
+                for geometry in geometry_collection.iter() {
+                    self.add_geometry(&GeometryCow::from(geometry));
+                }
+            }
+        }
+    }
+
+    fn add_polygon_ring(
+        &mut self,
+        linear_ring: &LineString<F>,
+        cw_left: CoordPos,
+        cw_right: CoordPos,
+    ) {
+        debug_assert!(linear_ring.is_closed());
+        if linear_ring.is_empty() {
+            return;
+        }
+
+        let mut coords: Vec<Coordinate<F>> = vec![];
+        // remove repeated coords
+        for coord in &linear_ring.0 {
+            if coords.last() != Some(coord) {
+                coords.push(*coord)
+            }
+        }
+
+        if coords.len() < 4 {
+            // TODO: we could return an Err here, but this has ramifications for how we can
+            // use this code in other operations - do we want all our methods, like `contains` to
+            // return a Result?
+            warn!("encountered invalid ring, which has undefined results");
+        }
+        let first_point = coords[0];
+
+        use crate::algorithm::winding_order::{Winding, WindingOrder};
+        let (left, right) = match linear_ring.winding_order() {
+            Some(WindingOrder::Clockwise) => (cw_left, cw_right),
+            Some(WindingOrder::CounterClockwise) => (cw_right, cw_left),
+            None => {
+                warn!("polygon had no winding order. Results are undefined.");
+                (cw_left, cw_right)
+            }
+        };
+
+        let edge = Edge::new(
+            coords,
+            Label::new(
+                self.arg_index,
+                TopologyPosition::area(CoordPos::OnBoundary, left, right),
+            ),
+        );
+        self.insert_edge(edge);
+
+        // insert the endpoint as a node, to mark that it is on the boundary
+        self.insert_point(self.arg_index, first_point, CoordPos::OnBoundary);
+    }
+
+    fn add_polygon(&mut self, polygon: &Polygon<F>) {
+        self.add_polygon_ring(polygon.exterior(), CoordPos::Outside, CoordPos::Inside);
+        // Holes are topologically labeled opposite to the shell, since
+        // the interior of the polygon lies on their opposite side
+        // (on the left, if the hole is oriented CW)
+        for hole in polygon.interiors() {
+            self.add_polygon_ring(hole, CoordPos::Inside, CoordPos::Outside)
+        }
+    }
+
+    fn add_line_string(&mut self, line_string: &LineString<F>) {
+        if line_string.is_empty() {
+            return;
+        }
+
+        let mut coords: Vec<Coordinate<F>> = vec![];
+        for coord in &line_string.0 {
+            if coords.last() != Some(coord) {
+                coords.push(*coord)
+            }
+        }
```

Comment:
is the conditional in the loop to avoid adding the endpoint? in most cases, the filter won't apply, so we could preallocate the vector on line 217 with `Vec::with_capacity`

Reply:
This was copied from JTS, but presumably there's no topological value in having redundant points adjacent in a vec.

I think you're right though - typically the filter won't apply.

I've done two things:
1. `Vec::with_capacity` here (and similarly in add_polygon)
2. `coords.shrink_to_fit()` in Edge construction, since these coord vec's never change len after this point.